### ISSUE TITLE
Mount won't recognize exfat file system automatically when no exfat volume label present

### ISSUE THREAD

**Issue by eZioPan** [relan/exfat#75]
# Test environment

CentOS 7 4.9.35-v7.1.el7 both on Raspberry Pi 2 and amd64 CPU
mkexfatfs 1.2.7
SanDisk USB3.0 64G Flash Disk

### Other helping system

Windows 10 x64

# Problem

for most situation, when I want to mount my usb flash disk, I need to use

`mount -t exfat /dev/sdb1 /mnt`

to mount my disk

but in one specific situation, I can use ( Step <1> below )

`mount /dev/sdb1 /mnt`

and the system will automatically recognize the exfat file system, then mount my flash disk

I'm confused about how this happened

**All report post here generate on Raspberry Pi 2**

## Step <1>

On Windows 10 x64, "Disk Manager", use "Delete Volume" and "New Simple Volume", use "exFAT" for "File System"

I plug my flash disk into linux

`mount /dev/sdb1 /mnt`

Successfully recognize file system

`dumpexfat /dev/sdb1`

Here is the output

Volume serial number      0x8e49c965
FS version                       1.0
Sector size                      512
Cluster size                  131072
Sectors count              122548224
Free sectors               122540544
Clusters count                478680
Free clusters                 478674
First sector                    2048
FAT first sector                2048
FAT sectors count               3840
First cluster sector            6144
Root directory cluster             4
Volume state                  0x0000
FATs count                         1
Drive number                    0x80
Allocated space                   0%

## Step <2>

After Step <1>

```
umount /dev/sdb1
mkfs.exfat /dev/sdb1
mount /dev/sdb1 /mnt
```

Get error

mount: /dev/sdb1: more filesystems detected. This should not happen,
       use -t <type> to explicitly specify the filesystem type or
       use wipefs(8) to clean up the device.

`dumpexfat /dev/sdb1`

Here is the output

Volume serial number      0x7c888b72
FS version                       1.0
Sector size                      512
Cluster size                  131072
Sectors count              122548224
Free sectors               122543360
Clusters count                478688
Free clusters                 478685
First sector                       0
FAT first sector                 128
FAT sectors count               3968
First cluster sector            4096
Root directory cluster             4
Volume state                  0x0000
FATs count                         1
Drive number                    0x80
Allocated space                   0%

## Step <3>

After Step <2>

```
umount /dev/sdb1
mkfs.exfat -p 2048 /dev/sdb1
mount /dev/sdb1 /mnt
```

Get error

mount: /dev/sdb1: more filesystems detected. This should not happen,
       use -t <type> to explicitly specify the filesystem type or
       use wipefs(8) to clean up the device.

`dumpexfat /dev/sdb1`

Here is the output

Volume serial number      0xa1731ffd
FS version                       1.0
Sector size                      512
Cluster size                  131072
Sectors count              122548224
Free sectors               122543360
Clusters count                478688
Free clusters                 478685
First sector                    2048
FAT first sector                 128
FAT sectors count               3968
First cluster sector            4096
Root directory cluster             4
Volume state                  0x0000
FATs count                         1
Drive number                    0x80
Allocated space                   0%

## Step <4>

On Windows 10 x64, "Disk Manager", use "Delete Volume" and "New Simple Volume", choose "Do not format this volume"
Then right click "Format" on my flash disk icon in "My Computer", choose "exFAT" and "Allocation unit size" choose "Default"

I plug my flash disk into linux

`mount /dev/sdb1 /mnt`

Get error

mount: /dev/sdb1: more filesystems detected. This should not happen,
       use -t <type> to explicitly specify the filesystem type or
       use wipefs(8) to clean up the device.

`dumpexfat /dev/sdb1`

Here is the output

Volume serial number      0x30a79536
FS version                       1.0
Sector size                      512
Cluster size                  131072
Sectors count              122548224
Free sectors               122540544
Clusters count                478680
Free clusters                 478674
First sector                    2048
FAT first sector                2048
FAT sectors count               3840
First cluster sector            6144
Root directory cluster             4
Volume state                  0x0000
FATs count                         1
Drive number                    0x80
Allocated space                   0%

## Step <5>

`fdisk /dev/sdb`

build a new DOS disklabel
make a new partition with default setting
change partition label to "HPFS/NTFS/exFAT"

`dumpexfat /dev/sdb1`

Here is the output

Volume serial number      0x30a79536
FS version                       1.0
Sector size                      512
Cluster size                  131072
Sectors count              122548224
Free sectors               122540544
Clusters count                478680
Free clusters                 478674
First sector                    2048
FAT first sector                2048
FAT sectors count               3840
First cluster sector            6144
Root directory cluster             4
Volume state                  0x0000
FATs count                         1
Drive number                    0x80
Allocated space                   0%

```
mkfs.exfat /dev/sdb1
mount /dev/sdb1 /mnt
```

Get error

mount: /dev/sdb1: more filesystems detected. This should not happen,
       use -t <type> to explicitly specify the filesystem type or
       use wipefs(8) to clean up the device.

`dumpexfat /dev/sdb1`

Here is the output

Volume serial number      0x5c77dcb9
FS version                       1.0
Sector size                      512
Cluster size                  131072
Sectors count              122550272
Free sectors               122545408
Clusters count                478696
Free clusters                 478693
First sector                       0
FAT first sector                 128
FAT sectors count               3968
First cluster sector            4096
Root directory cluster             4
Volume state                  0x0000
FATs count                         1
Drive number                    0x80
Allocated space                   0%

**Comment by relan** [comment]
Thanks for a detailed report!

After you run `umount /dev/sdb1`, is it really unmounted? What does `mount` say?

Before you run `mount /dev/sdb1 /mnt` that will fail, please try `blkid /dev/sdb1`. Does it detect FS properly?

**Comment by eZioPan** [comment]
Thanks for your reply!
It seems that I found why all this happened.
No matter how I format disk or create partition table, whether exfat partition has volume-name affects auto-recognition.
With default setting of mkfs.exfat, exfat partition won't get a volume-name, then when we mount it, we need to use mount -t exfat.
But with mkfs.exfat -n diskname, system will mount it without extra argument.

And I still wonder how this happened. ;-)

**Comment by relan** [comment]
> whether exfat partition has volume-name affects auto-recognition

Very good that you've managed to track down the issue. I'll have a look at util-linux code.

> could you please add a "remove volume label" at exfatlabel to clean volume label up?

It's already there, set empty label to remove it:

    exfatlabel /dev/sdb1 ''

I'll add this to the manpage.

**Comment by relan** [comment]
I can confirm that blkid (util-linux 2.23.2) from CentOS 7 fails to detect exFAT when FS is missing label. I've checked on the very same system blkid from vanilla util-linux 2.23.2 and it works fine. So it must one of 79 Red Hat's patches to util-linux that breaks exFAT detection. After all, always setting label is a viable workaround.

P.S. Added a note about how to remove volume name in e79bbfc02c0f0dbcca04e3a837af956e5ad3b2ac.
